### PR TITLE
PBC addition to hbond_analysis

### LINKED ISSUE #1188: In HydrogenBondAnalysis, include option to turn on periodic boundary conditions when searching for hydrogen bonds
HydrogenBondAnalysis searches for acceptors and donors within a certain distance cutoff but does not take into account the case when the acceptor and donor are located on different sides of the periodic box. This will happen when searching for hydrogen bonds in the water solvent.

To prevent these bonds from being missed, include an option to turn on periodic boundary conditions when doing the search for bonds. Default should be off because distance search with periodic boundary conditions is much slower and is only of interest in certain cases.

```

### Currently version of MDAnalysis:
(run `python -c "import MDAnalysis as mda; print(mda.__version__)"`)

### PR DESCRIPTION
**Pull request by navyakhare**
Fixes #1188 

Changes made in this Pull Request:
 -  Optionally applying PBC to the distance function, which calculates a single pairwise distance.

PR Checklist
------------
 - [Yes ] Tests
 - [No ] Docs
 - [No ] CHANGELOG updated
 - [Yes] Issue raised/referenced?

### CODE REVIEW

**Comment by navyakhare** [comment]
@richardjgowers  Please review the changes.

**Comment by kain88-de** [comment]
https://drive.google.com/file/d/1_QjyVswyay9N3uzVjizJebD5R24s9y_Y/view?usp=sharing

Here is an example of 4 atoms in a cubic box. One key point of the periodic boundary conditions is that we simulate an infinitely large system. This is happening because we do not only have the original box (blue in the picture) but an infinite number of identical images, the first surrounding layer is gray in the picture. In such a system it is not immediately clear what the distance between a particle A and B is. The minimal image convention (MIC) states that we should look at all distance of A to any B and use the minimum distance. In the picture that is marked as A to B'. 

Because of this simply packing atoms into the central box isn't enough to correctly calculate distance under PBC.  This picture also suggests test-cases. For two atoms on opposite edges of the box, the distance with PBC should be smaller than the normal Euclidean distance. For two atoms in the middle of the box there shouldn't be a difference.

**Comment by navyakhare** [comment]
@kain88-de thank you for the explanation I understand it much better now :smile:. 
@richardjgowers  test has been added.

File: testsuite/MDAnalysisTests/analysis/test_distances.py [hunk 1]
```diff
@@ -19,18 +19,21 @@
...
 # J. Comput. Chem. 32 (2011), 2319--2327, doi:10.1002/jcc.21787
 #
+
+
 from __future__ import print_function, absolute_import
 
 import pytest
 import scipy
 import scipy.spatial
+from math import *
```
Comment (richardjgowers) [inline testsuite/MDAnalysisTests/analysis/test_distances.py]: In general don't use `import *` ever, because it makes it unclear where a function came from.  Also, use `np.sqrt` not math

File: package/MDAnalysis/analysis/hbonds/hbond_analysis.py [hunk 2]
```diff
@@ -1030,11 +1031,12 @@ def _get_timestep():
     @staticmethod
     def calc_angle(d, h, a):
         """Calculate the angle (in degrees) between two atoms with H at apex."""
-        v1 = h.position - d.position
-        v2 = h.position - a.position
-        if np.all(v1 == v2):
-            return 0.0
-        return np.rad2deg(angle(v1, v2))
+        v1= d.position.reshape(1,len(d.position))
```
Comment (richardjgowers) [inline package/MDAnalysis/analysis/hbonds/hbond_analysis.py]: So to cast the array to be 2 dimensional you can do `v1[None, :]` which is a weird numpy trick to add a dimension

File: package/MDAnalysis/analysis/hbonds/hbond_analysis.py [hunk 3]
```diff
@@ -1030,11 +1031,12 @@ def _get_timestep():
     @staticmethod
     def calc_angle(d, h, a):
```
Comment (richardjgowers) [inline package/MDAnalysis/analysis/hbonds/hbond_analysis.py]: To make this handle PBC you need to add `box=None` as an optional kwarg to the signature here, then pass that to `calc_angles` (passing `None` to calc_angles is ok).  It's only possible to do PBC with the box information

File: testsuite/MDAnalysisTests/analysis/test_distances.py [hunk 4]
```diff
@@ -106,43 +108,76 @@ def test_box_sparse(self, coord, box, shape, res_pbc):
...
+            y= abs(ag.positions[i][1]-ag2.positions[i][1])
+            if y>4:
+                y=8-4
+            y=y*y
+            z= abs(ag.positions[i][2]-ag2.positions[i][2])
+            if z>4:
+                z=8-4
+            z=z*z
+            result=np.append(result,np.sqrt(x+y+z))
+
```
Comment (jmborr) [inline testsuite/MDAnalysisTests/analysis/test_distances.py]: You can substitute the loop for numpy array manipulation:
```
rp = np.abs(ag.positions - ag2.positions)
rp = np.where(rp > 4, 8.0 - rp, rp)
return np.sqrt(np.square(rp).sum(axis=1))
```

Comment (navyakhare) [reply testsuite/MDAnalysisTests/analysis/test_distances.py]: yes sure, I’ll remember to write neater code from next time.

**Comment by navyakhare** [comment]
@richardjgowers  @jmborr  , all changes made. Please review again.

File: testsuite/MDAnalysisTests/analysis/test_distances.py [hunk 5]
```diff
@@ -106,43 +108,64 @@ def test_box_sparse(self, coord, box, shape, res_pbc):
...
         )
 
+    @staticmethod
+    @pytest.fixture()
+    def expected_box(ag, ag2, box):
+
+        rp = np.abs(ag.positions - ag2.positions)
+        for i in range(len(rp)):
+            for j in range(len(rp[i])):
+                rp[i][j] = np.where(rp[i][j] > box[j]/2, box[j] - rp[i][j], rp[i][j])
```
Comment (jmborr) [inline testsuite/MDAnalysisTests/analysis/test_distances.py]: `np.where` doesn't make sense with scalars. Broadcast `box` to two dimensions and get rid of the double loop

```
box_2d = box[np.newaxis, 0:3]
rp = np.where(rp  > box_2d / 2, box_2d - rp, rp)
```

File: testsuite/MDAnalysisTests/analysis/test_hbonds.py [hunk 6]
```diff
@@ -28,9 +28,10 @@
 import pytest
 from MDAnalysis import SelectionError
 
-from numpy.testing import assert_equal, assert_array_equal, assert_almost_equal, assert_array_almost_equal
+from numpy.testing import assert_equal, assert_array_equal, assert_almost_equal, assert_array_almost_equal,assert_allclose
```
Comment (jmborr) [inline testsuite/MDAnalysisTests/analysis/test_hbonds.py]:
```
from numpy.testing import (assert_equal, assert_array_equal, assert_almost_equal, 
                           assert_array_almost_equal,assert_allclose)
```
or the thought police (a.k.a flake8) will complain

File: package/MDAnalysis/analysis/hbonds/hbond_analysis.py [hunk 7]
```diff
@@ -406,7 +407,7 @@ def __init__(self, universe, selection1='protein', selection2='all', selection1_
                  distance=3.0, angle=120.0,
                  forcefield='CHARMM27', donors=None, acceptors=None,
                  start=None, stop=None, step=None,
-                 debug=None, detect_hydrogens='distance', verbose=None):
+                 debug=None, detect_hydrogens='distance', verbose=None,pbc=True):
```
Comment (jmborr) [inline package/MDAnalysis/analysis/hbonds/hbond_analysis.py]: Include a description of this new optional argument in the the `Parameters` section of the class' docstring

File: package/MDAnalysis/analysis/hbonds/hbond_analysis.py [hunk 8]
```diff
@@ -582,6 +583,7 @@ def __init__(self, universe, selection1='protein', selection2='all', selection1_
         self.distance_type = distance_type  # note: everything except 'heavy' will give the default behavior
         self.angle = angle
         self.traj_slice = slice(start, stop, step)
+        self.pbc = pbc
```
Comment (jmborr) [inline package/MDAnalysis/analysis/hbonds/hbond_analysis.py]: Validate the dimensions when initializing `self.pbc`
```
self.pbc = pbc and flags['use_periodic_selections'] and all(self.u.dimensions[:3])
```

File: testsuite/MDAnalysisTests/analysis/test_hbonds.py (hunk 6 above)
Comment (kain88-de) [reply testsuite/MDAnalysisTests/analysis/test_hbonds.py]: This refers to [PEP8](https://www.python.org/dev/peps/pep-0008/) a common style guide for python. We try hard to follow it but do not enforce it. But please make the suggested change.

File: package/MDAnalysis/analysis/hbonds/hbond_analysis.py (hunk 8 above)
Comment (kain88-de) [reply package/MDAnalysis/analysis/hbonds/hbond_analysis.py]: The check seems false. If the `flags` variable is set false this would always evaluate to false and the pbc condition is never applied.

Comment (jmborr) [reply package/MDAnalysis/analysis/hbonds/hbond_analysis.py]: To be honest, I am not very familiar with setting the `flags` but I thought the default for `flags['use_periodic_selections']` was `True`, according to [selection.py](https://github.com/MDAnalysis/mdanalysis/blob/d5dec2b6f2be00fac35ee8f9be0106a244d6cc45/package/MDAnalysis/core/__init__.py#L270). Am I misreading this?

Comment (kain88-de) [reply package/MDAnalysis/analysis/hbonds/hbond_analysis.py]: The default can be `True` but the flags are changeable at runtime of the python session. You cannot rely on the fact that it is True. At the same time a setting of `pbc` should always overwrite the flag. The usual way of doing this is 

```python
def function(..., pbc=None):
    if pbc is None:
        pbc = flags['...']
```

This way we only use the flags if no value is provided for the function.

File: package/MDAnalysis/analysis/hbonds/hbond_analysis.py (hunk 7 above)
Comment (richardjgowers) [reply package/MDAnalysis/analysis/hbonds/hbond_analysis.py]: The default for pbc will need to be False otherwise this class will behave differently (with defaults) before and after this patch

File: package/MDAnalysis/analysis/hbonds/hbond_analysis.py (hunk 8 above)
Comment (richardjgowers) [reply package/MDAnalysis/analysis/hbonds/hbond_analysis.py]: flags are being removed soon, so I wouldn't add them anywhere any more

Comment (navyakhare) [reply package/MDAnalysis/analysis/hbonds/hbond_analysis.py]: so should I only validate the dimensions and not the flags ?

Comment (richardjgowers) [reply package/MDAnalysis/analysis/hbonds/hbond_analysis.py]: @navyakhare yeah ignore flags but if pbc=True do a sanity check like `all(u.dimensions[:3])` else some error about no dimension information

**Comment by navyakhare** [comment]
Sorry for the last few commits, I have reverted the state of the project back to 6c09ce5.

File: maintainer/install_hole.sh [hunk 9]
```diff
@@ -32,6 +32,12 @@ set -o errexit -o nounset
 
 SCRIPT=$(basename $0)
 
+<<<<<<< HEAD
```
Comment (richardjgowers) [inline maintainer/install_hole.sh]: this needs merging correctly, not sure why it caused a conflict

Comment (navyakhare) [reply maintainer/install_hole.sh]: I think I rebased already pushed commits while trying out something and hence it caused a conflict. It should be fine now that my local is back to the original state.

**Comment by navyakhare** [comment]
@richardjgowers  @kain88-de , any more changes needed or anything left to be implemented here?

**Comment by richardjgowers** [comment]
@navyakhare code looks good.  The diff is messy but I'll tidy that myself.  You need to add yourself to AUTHORS as it's your first contribution and then add a few lines in CHANGELOG (and add your name to the 17.1 authors) about this change

**Comment by navyakhare** [comment]
Sorry for the untidy diff and few unnecessary commits. This was the first time I contributed to an open source project and I learned a lot in the process. Thank you all for being so patient.
